chiselFrontend/src/main/scala/chisel3/core/Data.scala

```diff
@@ -258,9 +264,10 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
     _binding = Some(target)
   }
 
-  private[core] def topBindingOpt: Option[TopBinding] = _binding.map {
-    case ChildBinding(parent) => parent.topBinding
-    case bindingVal: TopBinding => bindingVal
+  private[core] def topBindingOpt: Option[TopBinding] = _binding.flatMap {
+    case ChildBinding(parent) => parent.topBindingOpt // Parent could be a SampleElement itself
+    case bindingVal: TopBinding => Some(bindingVal)
+    case _: SampleElementBinding[_] => None
```

Comment:
Consider making this assert out, since this really shouldn't happen ever, right? (sample element shouldn't be exposed to the user)

Reply:
`topBindingOpt` is called by [`Module` for allocating ids](https://github.com/freechipsproject/chisel3/blob/5c32cd87b85e2607e558f286cb1f8fd9604e4c83/chiselFrontend/src/main/scala/chisel3/core/UserModule.scala#L76). This change is explicitly making anything that has a `SampleElementBinding` parent (since you can have sample elements that are themselves aggregates) return None for it's top binding. The whole point of this change is to make sure that sample element isn't exposed to the user.

Reply:
It looks like `topBindingOpt` is used in a lot of places to indicate that something is an actual piece of hardware. I'm happy to wrap it in some way but in any case the `SampleElementBinding` will come up so it either needs to map to `None` in `topBindingOpt` or in some hypothetical new method that wraps `topBindingOpt` but only returns bindings that represent real hardware

Reply:
Oh I see.

Longer term, it might make sense to not have every HasId that gets instantiate try to add itself to the parent, since in some cases it doesn't make sense like for types (unbound), literals, and sample elements.

Reply:
Yeah I totally agree with the longer term vision. I'll add some comments to explain what `topBindingOpt` is and how it's used so that it's clear why `SampleElementBinding` returns `None`